### ISSUE TITLE
Kandinsky Pro training results are atrocious (Maybe fixed)

### ISSUE THREAD

**Issue by kabachuha** [bghira/SimpleTuner#2204]
Hello!

I'm trying to train a Kandinsky Pro lora using your trainer.

However, I'm stuck.

After a few epochs the video washes out and becomes a "mosaic". Example:

https://github.com/user-attachments/assets/905c844e-5943-46cd-b1d7-cd79e1291aac

I've made like 10 runs, maybe more, but nothing works. I've tried everything: turning on CREPA, turning it off; turning on/off layersync; turning on/off tread; turning every feature off except basic train; training on pure T2V; setting flow shift on/off (default and shift=5); using the default ["to_query", "to_key", "to_value"] or ["to_key", "to_query", "to_value", "self_attention.out_layer", "cross_attention.out_layer", "feed_forward.in_layer", "feed_forward.out_layer"] like in the official trainer.

The best result so far is:

https://github.com/user-attachments/assets/b70a602c-b341-4576-bbb4-c100a3474206

But after that it degenerated into the first video above.

My latest config is:

```json
{
    "caption_dropout_probability": 0.0,
    "checkpoint_epoch_interval": 2,
    "checkpoints_total_limit": 99,
    "compress_disk_cache": true,
    "data_backend_config": "/media/kabachuha/xiangliu/simpletunerx/kandinsly-multiresolution-crush.json",
    "disable_benchmark": true,
    "hub_model_id": "kandinsky5-video-19b-lora-test-crepa",
    "push_checkpoints_to_hub": false,
    "push_to_hub": false,
    "quantize_activations": false,
    "gradient_checkpointing": true,
    "musubi_blocks_to_swap": 20,
    "ramtorch": false,
    "ramtorch_vae": false,
    "peft_lora_target_modules": ["to_key", "to_query", "to_value", "self_attention.out_layer", "cross_attention.out_layer", "feed_forward.in_layer", "feed_forward.out_layer"],
    "learning_rate": 6e-5,
    "lora_rank": 16,
    "lora_type": "standard",
    "lora_format": "diffusers",
    "lr_scheduler": "polynomial",
    "lr_warmup_steps": 10,
    "max_grad_norm": 0.25,
    "num_train_epochs": 12,
    "minimum_image_size": 0,
    "mixed_precision": "bf16",
    "validation_disable": true,
    "offload_during_save": true,
    "base_model_precision": "int8-torchao",
    "model_family": "kandinsky5-video",
    "quantize_via": "cpu",
    "model_flavour": "i2v-pro-sft-5s",
    "pretrained_model_name_or_path": "/media/kabachuha/xiangliu/Kandinsky-5.0-I2V-Pro-sft-5s-Diffusers",
    "model_type": "lora",
    "optimizer": "optimi-lion",
    "flow_schedule_shift": 5.0,
    "output_dir": "/media/kabachuha/xiangliu/simpletunerx/kandinsky-models-shift",
    "report_to": "tensorboard",
    "resolution": 320,
    "resolution_type": "pixel_area",
    "resume_from_checkpoint": "latest",
    "seed": 6934,
    "vae_enable_patch_conv": false,
    "tracker_project_name": "kandinsky-lora-shift",
    "tracker_run_name": "kandinsky-training-shift",
    "train_batch_size": 1,
    "vae_batch_size": 1,
    "vae_enable_tiling": false,
    "vae_enable_slicing": false,
    "validation_guidance": 5.0,
    "validation_negative_prompt": "blurry, cropped, ugly",
    "validation_num_inference_steps": 20,
    "validation_num_video_frames": 73,
    "validation_prompt": "In the video, a miniature figure is presented. The figure is held in a person’s hands. The person then presses on the figure, causing a sq41sh squish effect. The person keeps pressing down on the figure, further showing the sq41sh squish effect.",
    "validation_prompt_library": false,
    "validation_resolution": "480x640",
    "validation_seed": 6934,
    "validation_steps": 50,
      "enable_group_offload": false,
      "group_offload_type": "block_level",
      "group_offload_blocks_per_group": 1,
      "group_offload_use_stream": true,
      "offload_during_startup": true,
  "attention_mechanism": "flash-attn-2",
  "crepa_enabled": true,
  "crepa_block_index": 8,
  "crepa_lambda": 0.02,
  "crepa_adjacent_distance": 1,
  "crepa_adjacent_tau": 1.0,
  "crepa_use_backbone_features": true,
  "crepa_teacher_block_index": 47,
  "crepa_drop_vae_encoder": false,
  "crepa_encoder": "dinov2_vitb14",
  "crepa_encoder_image_size": 544,
  "layersync_enabled": false,
  "layersync_student_block": 8,
  "layersync_teacher_block": 47,
  "layersync_lambda": 0.0012
}
```

I'm using the well-known "Crush" dataset at https://huggingface.co/datasets/GD-ML/Omni-VFX (crush subset) resized to 320x320x97f. And the dataset config is:

```json
[
  {
  "id": "video-480p-training",
  "type": "local",
  "dataset_type": "video",
  "crop": false,
  "instance_data_dir": "/media/kabachuha/xiangliu/dataset-crush-simpletuner-wen",
  "caption_strategy": "textfile",
  "resolution": 320,
  "resolution_type": "pixel_area",
  "minimum_image_size": 56,
  "maximum_image_size": 400,
  "target_downsample_size": 320,
  
  "minimum_aspect_ratio": 0.5,
  "maximum_aspect_ratio": 2.0,
  "conditioning_image_embeds": "conditioning-embeds-cache",
  "video": {
    "num_frames": 97,
    "frame_rate": 24,
    "target_resolution": 320,
    "resample_filter": "bicubic"
  },
  "conditioning_data": "kand-start-images",
  "cache_dir_vae": "/media/kabachuha/xiangliu/simpletunerx/kandinsky-crush-wen/vae-480p"
},{
  "id": "kand-start-images",
  "type": "local",
  "dataset_type": "conditioning",
  "crop": false,
  "instance_data_dir": "/media/kabachuha/xiangliu/dataset-crush-simpletuner-wen-images",
  "caption_strategy": "textfile",
  "resolution": 320,
  "resolution_type": "pixel_area",
  "minimum_image_size": 56,
  "maximum_image_size": 400,
  "target_downsample_size": 320,
  "minimum_aspect_ratio": 0.5,
  "maximum_aspect_ratio": 2.0,
    "video": {
    "num_frames": 97,
    "frame_rate": 24,
    "target_resolution": 320,
    "resample_filter": "bicubic"
  },
  "cache_dir_vae": "/media/kabachuha/xiangliu/simpletunerx/kandinsky-crush-wen/vae-480p-kand-start"
},
  {
    "id": "text-embed-cache",
    "dataset_type": "text_embeds",
    "default": true,
    "type": "local",
    "cache_dir": "/media/kabachuha/xiangliu/simpletunerx/kandinsky-crush-wen/text"
  }, { "id": "conditioning-embeds-cache", "type": "local", "dataset_type": "conditioning_image_embeds", "disabled": false, "cache_dir": "/media/kabachuha/xiangliu/simpletunerx/kandinsky-crush-wen/cond" }
]
```

I've been trying to train it multiple days with various combinations, but no luck. Could you help me either with the settings or to make sure Kandinsky 5 Pro LoRA training works correctly?

Also, I'm using a hack with the conditioning dataset. Because of the training demanding the "conditioning" dataset and failing to parse videofiles on the first run, I have to create an images-filled folder which I use as conditioning for the first run and after it crashes I redirect it to videos-only filled folder. I don't think it's an intended behavior. Anyway, it seems to be an orthogonal problem, because the text2video train fails as well.

**Comment by bghira** [comment]
that does sound frustrating. i see that you're using optimi-lion, have you tried adamw_bf16?

the pure bf16 training that simpletuner does is sensitive to small gradients simply vanishing. i'm not sure that's what's happening here, but i've verified the VAE normalisation and scaling is correct, so something about the int8-torchao or bf16, or optimiser, could be causing further grief.

**Comment by bghira** [comment]
```
      "group_offload_use_stream": true,
```

there may also be a bug with this (but i see it's not enabled here anyway)

**Comment by kabachuha** [comment]
Thank you!

Changing the optimizer from lion to adamw_bf16, removing the parts about group offload and setting max_grad_norm to 1.0 has fully fixed this coloration / mosaic problem even at the end of the training run! (the mosaic used to express itself somewhere in the middle)

https://github.com/user-attachments/assets/37f22d63-e668-438a-befd-d093dc60e438

However, please don't close this issue yet.  I need to do additional / longer tests to make sure it didn't express itself yet because of underfitting.

**Comment by bghira** [comment]
ok, Lion is probably the issue entirely I'd assume.. it's bad with models that are not rotation-invariant, and i wonder whether Kandinsky5 Video suffers the same. the Lion-K optimiser was introduced recently to overcome that:

> The territorial lion safeguards its territory from outside attackers, especially nomadic lions. This process is called territorial defense. It protects the cubs till they become--

oops, wrong kind of Lion!

> https://arxiv.org/abs/2310.05898

this paper focuses briefly on Conv2D but it's not so much related to this operation as the general sign(-) factor that's applied, because it treats every parameter as if they are independent. this works for Linear layers in Transformer models, but Conv parameters are group-aware. 

when the Conv is updated by Lion, it doesn't have spatial awareness to account for the required 180 degrees of symmetry or 90 degree orbit. 

symmetrical requirement means once the transform occurs, it is defined as symmetrical if the outcome is invariant to the rotation:

```
Original kernel:      Rotated 90° clockwise:
[a b c]               [g d a]
[d e f]      →        [h e b]
[g h i]               [i f c]
```

the orbital requirement looks like this:

```
[. b .]     b → f → h → d → b  (4 positions)
[d . f]
[. h .]
```

with the centre being its own orbit:

```
[. . .]
[. e .]     e → e  (1 position, fixed point)
[. . .]
```

rotational invariance requires that `a=c` and since Lion is treaing parameters independently, we end up with a collapse condition.

one of the OneTrainer developers introduced Lion-K to that project with this proposed theoretical basis of why Lion performs poorly in Unet models. the lion-k paper doesn't get into these architecture-specific issues, so, the conclusion is admittedly still lacking theoretical grounding beyond this simple demonstration.

i suppose i should update the examples to have better optimiser selection for these architectures.

**Comment by kabachuha** [comment]
Thank you for the explaination! It seems to be Lion indeed

Now (even with TREAD and CREPA (backbone)) it works perfectly after >1.5 times the training time. (~ 3 hours of training)

https://github.com/user-attachments/assets/7ef95bfc-c3e2-4d1d-9cc4-739a25cb1f64

https://github.com/user-attachments/assets/1b165b34-e3ef-4b23-ad7c-96ba19fc452e

Closing this issue now!

Also, thank you for CREPA. Without it,. Kohya-ss trainer produced subpar results even after > 8 hours.

Kohya (musubi) result below:

https://github.com/user-attachments/assets/a7448164-8ca5-4b5b-9b37-70582e984527

**Comment by bghira** [comment]
great, it seems like my fallback theory of the Pro model's reinforcement learning disrupting downstreaming finetuning is just a moot point to explore.
